expo/expo
# [expo-updates][ios] track build configuration data

# Why

 The build data stored by the configuration is subject to change when a user updates the binary.
 
 This can lead to inconsistent update loading behavior, for example if a user were to publish to channel A after a build with channel B was time stamped, then when the client downloaded build B they would still get update A, instead of the expected build B embedded update:
 
<img width="546" alt="Screen Shot 2021-11-04 at 11 02 58 AM" src="https://user-images.githubusercontent.com/1220444/140403789-ec128421-ada0-4a1a-93bc-3104e74de18a.png">
 
 c.f. https://github.com/expo/expo/issues/14372
 
 
 
# How

Introduce a singleton that wipes the updates when any of the tracked build data changes. This leaves the user in the same situation as a fresh install.
 
While the original report is for `releaseChannel` and also applies to `requestHeaders.channel`, a few extra fields are included for safety after discussion with @esamelson 

# Test Plan
(as always with expo-updates, make sure you are building for release and not debug)

1. create build
2. publish update
3. load update
4.  change one of the fields tracked by `buildData` such as `releaseChannel`
5. edit the commitTime in `scripts/createManifest.js` back before the time of the initial build and publish.
6. rebuild  and load without deleting the previous

Since this is back in time, if the updates are not cleared, the published update will load. With this fix applied the embedded update from step 5 should load.



# Checklist
